src/lib/pickles/wrap_verifier.ml

```diff
@@ -198,55 +198,99 @@ struct
-                    (b, x)
-                | Plonk_types.Opt.Some x ->
-                    (Boolean.true_, x)
-              in
-              let b1, x1 = to_maybe x in
-              let b2, x2 = to_maybe y in
-              Plonk_types.Opt.Maybe (f_bool b1 b2, f x1 x2) )
-    in
     fun bs keys ->
       let open Field in
       Vector.map2
         (bs :> (Boolean.var, n) Vector.t)
         keys
         ~f:(fun b key ->
-          map key
+          Plonk_verification_key_evals.Step.map key
             ~f:(fun g -> Double.map g ~f:(( * ) (b :> t)))
-            ~f_bool:(fun b_inner -> Boolean.(b &&& b_inner)) )
+            ~f_opt:(function
+              (* Here, we split the 3 variants into 3 separate accumulators. This
+                 allows us to only compute the 'maybe' flag when we need to, and
+                 allows us to fall back to the basically-free `None` when a
+                 feature is entirely unused, or to the less expensive `Some` if
+                 it is used for every circuit.
+                 In particular, it is important that we generate exactly `None`
+                 when none of the optional gates are used, otherwise we will
+                 change the serialization of the protocol circuits.
+              *)
+              | Plonk_types.Opt.None ->
+                  ([], [], [ b ])
+              | Plonk_types.Opt.Maybe (b_x, x) ->
+                  ([], [ (b, b_x, x) ], [])
+              | Plonk_types.Opt.Some x ->
+                  ([ (b, x) ], [], []) ) )
       |> Vector.reduce_exn
            ~f:
-             (map2 ~f:(Double.map2 ~f:( + ))
-                ~f_bool:(fun (b1 : Boolean.var) (b2 : Boolean.var) ->
-                  Boolean.Unsafe.of_cvar
-                    Field.(add (b1 :> Field.t) (b2 :> Field.t)) ) )
-      |> map
-           ~f:(fun g -> Double.map ~f:(Util.seal (module Impl)) g)
-           ~f_bool:(fun (b : Boolean.var) ->
-             Boolean.Unsafe.of_cvar (Util.seal (module Impl) (b :> t)) )
+             (Plonk_verification_key_evals.Step.map2 ~f:(Double.map2 ~f:( + ))
+                ~f_opt:(fun (yes_1, maybe_1, no_1) (yes_2, maybe_2, no_2) ->
+                  (yes_1 @ yes_2, maybe_1 @ maybe_2, no_1 @ no_2) ) )
+      |> Plonk_verification_key_evals.Step.map ~f:Fn.id ~f_opt:(function
+           | [], [], _nones ->
+               (* We only have `None`s, so we can emit exactly `None` without
+                  further computation.
+               *)
+               Plonk_types.Opt.None
+           | somes, [], [] ->
+               (* Special case: we don't need to compute the 'maybe' bool
+                  because we know statically that all entries are `Some`.
+               *)
+               let sum =
+                 somes
+                 |> List.map ~f:(fun ((b : Boolean.var), g) ->
+                        Double.map g ~f:(( * ) (b :> t)) )
+                 |> List.reduce_exn ~f:(Double.map2 ~f:( + ))
```

Comment:
Probably could fuse the 2 iterations. But the use of `reduce_exn` would make it less nice.